script/src/main/java/com/focusit/script/ScriptEngine.java

```diff
@@ -1,115 +1,103 @@
-    private static final ThreadLocal<HashMap<String, Script>> threadBindedScripts = new ThreadLocal<>();
+    private static final ScriptEngine INSTANCE = new ScriptEngine();
     private static final Logger LOG = LoggerFactory.getLogger(ScriptEngine.class);
-    // General purpose script storage.
-    // Not Thread Safe, because compiled script has writable bindings. So one thread can easily change other one's bindings
-    private final ConcurrentHashMap<String, Script> generalScripts = new ConcurrentHashMap<>();
-    private final Script NO_SCRIPT = new Script()
+    private static final ConcurrentHashMap<String, Class<? extends Script>> SCRIPT_CLASSES = new ConcurrentHashMap<>();
+    private static Boolean initialized = false;
+    private static GroovyClassLoader groovyClassLoader;
+
+    private ScriptEngine() {}
+
+    private static void throwIfNotInitialized()
     {
-        @Override
-        public Object run()
+        if (!initialized)
         {
-            return null;
+            throw new IllegalStateException("ScriptEngine wasn't initialized");
         }
-    };
-    private volatile ClassLoader loader;
-    private GroovyShell shell;
+    }
 
-    public ScriptEngine(ClassLoader classLoader)
+    private static ScriptEngine getInstance()
     {
-        this.loader = classLoader;
-        shell = new GroovyShell(loader);
+        throwIfNotInitialized();
+        return INSTANCE;
     }
 
-    public ClassLoader getClassLoader()
+    public static Script getScript(String scriptBody)
     {
-        return this.loader;
+        return getInstance().getScriptInternal(scriptBody);
     }
 
-    @Nullable
-    public Script getScript(String script)
+    public static void init(ClassLoader classLoader)
     {
-        if (script == null)
+        if (initialized)
         {
-            return null;
+            return;
         }
+        groovyClassLoader = new GroovyClassLoader(classLoader);
+        initialized = true;
+    }
 
-        if (script.isEmpty())
+    public static ClassLoader getClassLoader()
+    {
+        return getInstance().getClassLoaderInternal();
+    }
+
+    private Script getScriptInternal(String scriptBody)
+    {
+        if (StringUtils.isNullOrEmptyOrWhiteSpace(scriptBody))
         {
             return null;
         }
 
-        Script result = generalScripts.get(script);
-
-        if (result == null)
+        Class<? extends Script> clazz = SCRIPT_CLASSES.get(scriptBody);
+        if (clazz == null)
         {
-            if (script.trim().length() > 0)
+            synchronized (SCRIPT_CLASSES)
             {
-                result = shell.parse(script);
+                if ((clazz = SCRIPT_CLASSES.get(scriptBody)) == null)
+                {
+                    Class<? extends Script> tempClazz = groovyClassLoader.parseClass(scriptBody);
+                    SCRIPT_CLASSES.put(scriptBody, tempClazz);
+                    clazz = tempClazz;
+                }
```

Comment:
this whole block could be replaced by SCRIPT_CLASSES.computeIfAbsent
FYI ConcurrentHashMap synchronizes itself internally, there is no need of extertnal synchronization